rcl_action/src/rcl_action/action_server.c

```diff
@@ -170,9 +170,12 @@ rcl_action_server_init(
   PUBLISHER_INIT(feedback);
   PUBLISHER_INIT(status);
 
-  // Initialize Timer
+  // Copy clock
+  action_server->impl->clock = *clock;
+
```

Comment:
This is a question to make sure I understand.  Is the purpose of copying the clock here so that if `rcl_timer_init` fails, `rcl_action_server_fini` in the `fail` block below has a valid pointer to cleanup?

Reply:
Pretty much, basically. This one is a bit of a rabbit hole, so bare with me. In `rcl_timer_init`, it copies the pointer to the `clock` pointer passed to it. In `rcl_timer_fini` (called in `rcl_action_server_fini` line 235), it will try to access fields of that pointer by calling `rcutils_allocator_is_valid`, `rcl_clock_remove_jump_callback` and checking its type on the same clock pointer. Since ownership of the `clock` pointer is not supposed to be transferred in this function (rclcpp::Node still owns it), then it should be using the pointer to action_server`s clock instead during `rcl_timer_init`.

This was only found when running valgrind on rclcpp`s `test_server`. Basically, rclcpp::Node was destroying the clock before `rcl_timer_fini` was called in this action_server. Trackback below to one of the memory read errors without this fix.

```
==2946848== Invalid read of size 8
==2946848==    at 0x5896463: rcutils_allocator_is_valid (allocator.c:96)
==2946848==    by 0x57DB111: rcl_clock_remove_jump_callback (time.c:431)
==2946848==    by 0x57DCD41: rcl_timer_fini (timer.c:217)
==2946848==    by 0x49596D6: rcl_action_server_fini (action_server.c:235)
==2946848==    by 0x48E8666: rclcpp_action::ServerBase::ServerBase(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeClockInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeLoggingInterface>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rosidl_action_type_support_t const*, rcl_action_server_options_t const&)::{lambda(rcl_action_server_t*)#1}::operator()(rcl_action_server_t*) const (server.cpp:90)
==2946848==    by 0x48F167F: std::_Sp_counted_deleter<rcl_action_server_t*, rclcpp_action::ServerBase::ServerBase(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeClockInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeLoggingInterface>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rosidl_action_type_support_t const*, rcl_action_server_options_t const&)::{lambda(rcl_action_server_t*)#1}, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() (shared_ptr_base.h:471)
==2946848==    by 0x3432B5: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (shared_ptr_base.h:155)
==2946848==    by 0x3334E2: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (shared_ptr_base.h:730)
==2946848==    by 0x48F21CD: std::__shared_ptr<rcl_action_server_t, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (shared_ptr_base.h:1169)
==2946848==    by 0x48F2211: std::shared_ptr<rcl_action_server_t>::~shared_ptr() (shared_ptr.h:103)
==2946848==    by 0x48F5CF3: rclcpp_action::ServerBaseImpl::~ServerBaseImpl() (server.cpp:35)
==2946848==    by 0x48F5D53: std::default_delete<rclcpp_action::ServerBaseImpl>::operator()(rclcpp_action::ServerBaseImpl*) const (unique_ptr.h:81)


==2946848==  Address 0xa0d7738 is 40 bytes inside a block of size 160 free'd
==2946848==    at 0x483F1CF: operator delete(void*, unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2946848==    by 0x50F4571: std::_Sp_counted_ptr<rclcpp::Clock::Impl*, (__gnu_cxx::_Lock_policy)2>::_M_dispose() (shared_ptr_base.h:377)
==2946848==    by 0x3432B5: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (shared_ptr_base.h:155)
==2946848==    by 0x3334E2: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (shared_ptr_base.h:730)
==2946848==    by 0x50F343C: std::__shared_ptr<rclcpp::Clock::Impl, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (shared_ptr_base.h:1169)
==2946848==    by 0x50F346E: std::shared_ptr<rclcpp::Clock::Impl>::~shared_ptr() (shared_ptr.h:103)
==2946848==    by 0x50F1212: rclcpp::Clock::~Clock() (clock.cpp:64)
==2946848==    by 0x44E865: void __gnu_cxx::new_allocator<rclcpp::Clock>::destroy<rclcpp::Clock>(rclcpp::Clock*) (new_allocator.h:153)
==2946848==    by 0x44637B: void std::allocator_traits<std::allocator<rclcpp::Clock> >::destroy<rclcpp::Clock>(std::allocator<rclcpp::Clock>&, rclcpp::Clock*) (alloc_traits.h:497)
==2946848==    by 0x4309C0: std::_Sp_counted_ptr_inplace<rclcpp::Clock, std::allocator<rclcpp::Clock>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() (shared_ptr_base.h:557)
==2946848==    by 0x3432B5: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (shared_ptr_base.h:155)
==2946848==    by 0x3334E2: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (shared_ptr_base.h:730)
```

Reply:
Thanks, great explanation.  That's really helpful.